### ISSUE TITLE
Custom kernel for faster asynchronous mask generation

### ISSUE THREAD

**Issue by jcrozum** [jcrozum/cubewalkers#14]
The asynchronous mask generation has been sped up substantially, but it is still far from optimal. A custom kernel is probably needed to do it properly.

**Comment by troonmel** [comment]
I tried writing a code using raw kernel, but it is still slower.

asynchronous        :    CPU:  478.427 us   +/-51.972 (min:  414.300 / max: 1842.300) us     GPU-0:  526.902 us   +/-54.446 (min:  361.472 / max: 1899.520) us
asynchronous_new    :    CPU:  632.624 us   +/-81.736 (min:  426.800 / max: 5656.900) us     GPU-0: 1020.555 us   +/-140.970 (min:  518.112 / max: 5899.264) us

```
asynchronous_kernel = cp.RawKernel(r'''
extern "C" __global__
void asynchronous(const int* x1, int* z, int N, int W) {
    int n_reserved = blockDim.x * blockIdx.x + threadIdx.x;
    int w_reserved = blockDim.y * blockIdx.y + threadIdx.y;
    if(n_reserved < N && w_reserved < W){
    if(x1[w_reserved]==n_reserved){
    int a_reserved = n_reserved * W + w_reserved;
    z[a_reserved] = 1;}}
}
''', 'asynchronous')

def asynchronous_new(t: int, n: int, w: int, a: cp.ndarray) -> cp.ndarray:
    x1 = cp.random.randint(low=0, high=n, size=(w,), dtype=cp.int_)
    z = cp.zeros((n, w), dtype=cp.int_)
    asynchronous_kernel((n//32+1,w//32+1), (32,32), (x1, z, n, w))

    return z.astype(cp.float32)
```

**Comment by jcrozum** [comment]
Are you sure you're testing with the same parameters? When I test it on my end, this is what I get as output:
```
old: <lambda>            :    CPU:  953.375 us   +/-269.429 (min:  847.632 / max: 2314.934) us     GPU-0: 1029.836 us   +/-299.623 (min:  895.072 / max: 3457.024) us
new: <lambda>            :    CPU:  183.373 us   +/-62.497 (min:  166.880 / max:  594.182) us     GPU-0:  304.614 us   +/-60.269 (min:  224.064 / max: 1017.856) us
```

Your new version is much faster. Here I was using n=w=1000 and using 200 trials.

**Comment by troonmel** [comment]
I see that it scales better with N. new version is faster for N=1000, but slower for N=100.

**Comment by jcrozum** [comment]
That kind of makes sense; higher N means the old version is more likely to trigger the while loop.

**Comment by jcrozum** [comment]
Try
```python
x1 = cp.floor(cp.random.random(size=(w,))*n)
```
This was faster in all my tests.

**Comment by troonmel** [comment]
Makes sense as randint is very slow. I think it is taking time to make sure that every integer is chosen with exact equal probability. The above method may have very small differences depending on how many numbers lie in each interval (which I think is negligible considering that random uses float64 as default).

**Comment by jcrozum** [comment]
Resolved in #16
